File: api/src/main/java/fr/huiitre/tools/modules/riot/valorant/application/user/usecase/ValorantWatchlistNotifier.java
package fr.huiitre.tools.modules.riot.valorant.application.user.usecase;

import fr.huiitre.tools.modules.core.notification.application.event.NotificationEvent;
import fr.huiitre.tools.modules.core.notification.domain.entity.NotificationType;
import fr.huiitre.tools.modules.riot.valorant.application.catalog.ports.ValorantSkinRepository;
import fr.huiitre.tools.modules.riot.valorant.application.catalog.ports.ValorantStorePort;
import fr.huiitre.tools.modules.riot.valorant.application.core.ports.ValorantVersionProvider;
import fr.huiitre.tools.modules.riot.valorant.application.core.ports.ValorantAuthRepository;
import fr.huiitre.tools.modules.riot.valorant.application.core.usecase.ValorantAuthService;
import fr.huiitre.tools.modules.riot.valorant.application.skin.view.ValorantSkinView;
import fr.huiitre.tools.modules.riot.valorant.application.user.ports.ValorantStoreHistoryRepository;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.context.ApplicationEventPublisher;
import org.springframework.stereotype.Service;

import java.time.Instant;
import java.time.LocalDate;
import java.time.ZoneOffset;
import java.util.ArrayList;
import java.util.List;
import java.util.Optional;
import java.util.UUID;
import java.util.stream.Collectors;

@Service
public class ValorantWatchlistNotifier {

    private static final Logger log = LoggerFactory.getLogger(ValorantWatchlistNotifier.class);

    private final ValorantAuthService valorantAuthService;
    private final ValorantAuthRepository valorantAuthRepository;
    private final ValorantStorePort valorantStorePort;
    private final ValorantVersionProvider versionProvider;
    private final ValorantSkinRepository skinRepository;
    private final ValorantStoreHistoryRepository storeHistoryRepository;
    private final ApplicationEventPublisher eventPublisher;

    public ValorantWatchlistNotifier(ValorantAuthService valorantAuthService,
                                   ValorantAuthRepository valorantAuthRepository,
                                   ValorantStorePort valorantStorePort,
                                   ValorantVersionProvider versionProvider,
                                   ValorantSkinRepository skinRepository,
                                   ValorantStoreHistoryRepository storeHistoryRepository,
                                   ApplicationEventPublisher eventPublisher) {
        this.valorantAuthService = valorantAuthService;
        this.valorantAuthRepository = valorantAuthRepository;
        this.valorantStorePort = valorantStorePort;
        this.versionProvider = versionProvider;
        this.skinRepository = skinRepository;
        this.storeHistoryRepository = storeHistoryRepository;
        this.eventPublisher = eventPublisher;
    }

    public void processAllAccounts() {
        // On récupère uniquement les comptes Valorant liés avec des credentials persistés
        List<Long> accountIds = valorantAuthRepository.findAllAccountIds();
        log.info("Starting Valorant Watchlist & History sync for {} accounts", accountIds.size());

        for (Long accountId : accountIds) {
            try {
                processAccount(accountId);
            } catch (Exception e) {
                log.error("Error processing Valorant sync for account {}: {}", accountId, e.getMessage());
            }
        }
    }

    private void processAccount(Long accountId) {
        Optional<ValorantAuthRepository.ValorantAuthData> authDataOpt = valorantAuthRepository.findById(accountId);
        if (authDataOpt.isEmpty()) return;

        ValorantAuthRepository.ValorantAuthData authData = authDataOpt.get();
        String accessToken = valorantAuthService.getOrRefreshAccessToken(accountId);
        String entitlementsToken = valorantStorePort.fetchEntitlementsToken(accessToken);
        String clientVersion = versionProvider.getVersion().get("riotClientVersion").toString();

        ValorantStorePort.RawStorefront raw = valorantStorePort.fetchStorefront(
                authData.puuid(), authData.region(), accessToken, entitlementsToken, clientVersion
        );

        // 1. Archivage automatique dans l'historique
        LocalDate shopDate = calculateShopDate(raw.singleItemOffersRemainingDurationInSeconds());
        List<Long> skinDbIds = new ArrayList<>();

        for (ValorantStorePort.RawOffer offer : raw.singleItemOffers()) {
            skinRepository.findByLevelAssetId(UUID.fromString(offer.itemId()), accountId)
                    .ifPresent(skin -> {
                        skinDbIds.add(skin.id());
                        if (!storeHistoryRepository.existsByAccountIdAndSkinIdAndDate(accountId, skin.id(), shopDate)) {
                            storeHistoryRepository.add(accountId, skin.id(), shopDate);
                        }
                    });
        }

        // 2. Vérification Watchlist & Notification
        List<ValorantSkinView> watchlist = skinRepository.findAllWatchedByAccountId(accountId);
        List<ValorantSkinView> matches = watchlist.stream()
                .filter(w -> skinDbIds.contains(w.id()))
                .collect(Collectors.toList());

        if (!matches.isEmpty()) {
            sendNotification(authData.userId(), accountId, matches);
        }
    }

    private LocalDate calculateShopDate(long remainingSeconds) {
        // On se cale sur le milieu de la rotation pour stabiliser la date (comme au front)
        // Expiration - 12h
        return Instant.now()
                .plusSeconds(remainingSeconds)
                .minusSeconds(43200) // 12h
                .atZone(ZoneOffset.UTC)
                .toLocalDate();
    }

    private void sendNotification(long userId, Long accountId, List<ValorantSkinView> matches) {
        String accountLabel = valorantAuthRepository.findAllByUserId(userId).stream()
                .filter(a -> a.id() == accountId)
                .findFirst()
                .map(a -> a.label() != null ? a.label() : (a.gameName() != null ? a.gameName() + "#" + a.tagLine() : null))
                .orElse(null);
        String accountSuffix = accountLabel != null ? " (" + accountLabel + ")" : "";

        String title = "Valorant Shop - Skins disponibles !" + accountSuffix;
        String body;

        if (matches.size() == 1) {
            body = "Le skin \"" + matches.get(0).name() + "\" est disponible dans ta boutique !";
        } else {
            String names = matches.stream()
                    .map(s -> "- " + s.name())
                    .collect(Collectors.joining("\n"));
            body = "Plusieurs skins de ta watchlist sont disponibles :\n" + names;
        }

        eventPublisher.publishEvent(new NotificationEvent(
                title,
                body,
                NotificationType.SUCCESS,
                userId,
                null,
                null,
                null,
                "{\"route\": \"valorant-shop\"}"
        ));
    }
}

File: api/Modules/Riot/Valorant/Application/User/Services/ValorantWatchlistNotifier.cs
using Tools.Api.Modules.Core.Notifications.Application;
using Tools.Api.Modules.Core.Notifications.Application.Services;
using Tools.Api.Modules.Riot.Valorant.Application.Catalog.Ports;
using Tools.Api.Modules.Riot.Valorant.Application.Core.Ports;
using Tools.Api.Modules.Riot.Valorant.Application.Core.Services;
using Tools.Api.Modules.Riot.Valorant.Application.Skin.Views;
using Tools.Api.Modules.Riot.Valorant.Application.User.Ports;

namespace Tools.Api.Modules.Riot.Valorant.Application.User.Services;

// Passe sur tous les comptes liés : archive la boutique du jour et prévient l'utilisateur quand
// un skin de sa liste de suivi y est en vente.
//
// **Ce n'est pas un use case sécurisé, et ça ne doit jamais le devenir** : le planificateur
// l'appelle sur un thread de fond, où aucun utilisateur n'est identifié — hériter de
// SecuredUseCase ferait échouer l'autorisation dès la construction. Le déclenchement manuel passe
// par TriggerValorantWatchlistSyncUseCase, qui porte le contrôle d'accès.
public sealed class ValorantWatchlistNotifier(
    ValorantAuthService valorantAuthService,
    IValorantAuthRepository valorantAuthRepository,
    IValorantStorePort valorantStorePort,
    IValorantVersionProvider versionProvider,
    IValorantSkinRepository skinRepository,
    IValorantStoreHistoryRepository storeHistoryRepository,
    NotificationService notificationService,
    ILogger<ValorantWatchlistNotifier> logger)
{
    private const string NotificationMetadata = """{"route": "valorant-shop"}""";

    // La rotation dure 24 h : on date l'archive de son milieu, pour que la même journée de
    // boutique porte la même date d'un bout à l'autre. Le front applique le même calcul.
    private static readonly TimeSpan HalfRotation = TimeSpan.FromHours(12);

    public async Task ProcessAllAccounts()
    {
        var accountIds = await valorantAuthRepository.FindAllAccountIds();

        logger.LogInformation(
            "Synchronisation Valorant (historique + liste de suivi) pour {AccountCount} compte(s)",
            accountIds.Count);

        foreach (var accountId in accountIds)
        {
            try
            {
                await ProcessAccount(accountId);
            }
            catch (Exception exception)
            {
                // Un compte en échec — jeton périmé, Riot injoignable — ne doit pas priver les
                // autres de leur passe.
                logger.LogError(
                    exception,
                    "Échec de la synchronisation Valorant du compte {AccountId}",
                    accountId);
            }
        }
    }

    private async Task ProcessAccount(long accountId)
    {
        if (await valorantAuthRepository.FindById(accountId) is not { } authData)
        {
            return;
        }

        var accessToken = await valorantAuthService.GetOrRefreshAccessToken(accountId);
        var entitlementsToken = await valorantStorePort.FetchEntitlementsToken(accessToken);
        var clientVersion = await versionProvider.GetRiotClientVersion();

        var raw = await valorantStorePort.FetchStorefront(
            authData.Puuid, authData.Region, accessToken, entitlementsToken, clientVersion);

        var shopDate = ShopDateOf(raw.SingleItemOffersRemainingDurationInSeconds);
        var skinIdsInShop = new List<long>();

        foreach (var offer in raw.SingleItemOffers)
        {
            if (!Guid.TryParse(offer.ItemId, out var levelAssetId))
            {
                continue;
            }

            if (await skinRepository.FindByLevelAssetId(levelAssetId, accountId) is not { } skin)
            {
                continue;
            }

            skinIdsInShop.Add(skin.Id);

            if (!await storeHistoryRepository.ExistsByAccountIdAndSkinIdAndDate(accountId, skin.Id, shopDate))
            {
                await storeHistoryRepository.Add(accountId, skin.Id, shopDate);
            }
        }

        var watched = await skinRepository.FindAllWatchedByAccountId(accountId);
        var matches = watched.Where(skin => skinIdsInShop.Contains(skin.Id)).ToList();

        if (matches.Count > 0)
        {
            await Notify(authData.UserId, accountId, matches);
        }
    }

    private static DateOnly ShopDateOf(long remainingSeconds)
    {
        return DateOnly.FromDateTime(
            DateTime.UtcNow.AddSeconds(remainingSeconds).Subtract(HalfRotation));
    }

    private async Task Notify(long userId, long accountId, List<ValorantSkinView> matches)
    {
        var accountLabel = await AccountLabelOf(userId, accountId);
        var suffix = accountLabel is null ? string.Empty : $" ({accountLabel})";

        var body = matches.Count == 1
            ? $"Le skin \"{matches[0].Name}\" est disponible dans ta boutique !"
            : "Plusieurs skins de ta liste de suivi sont disponibles :\n"
              + string.Join("\n", matches.Select(skin => $"- {skin.Name}"));

        await notificationService.Send(new SendNotificationCommand(
            $"Valorant Shop - Skins disponibles !{suffix}",
            body,
            NotificationType.Success,
            TargetUserId: userId,
            Metadata: NotificationMetadata));
    }

    // Le libellé donné par l'utilisateur, sinon son Riot ID : avec plusieurs comptes liés, une
    // notification sans nom ne dit pas lequel est concerné.
    private async Task<string?> AccountLabelOf(long userId, long accountId)
    {
        var accounts = await valorantAuthRepository.FindAllByUserId(userId);

        if (accounts.FirstOrDefault(account => account.Id == accountId) is not { } account)
        {
            return null;
        }

        if (!string.IsNullOrWhiteSpace(account.Label))
        {
            return account.Label;
        }

        return account.GameName is null ? null : $"{account.GameName}#{account.TagLine}";
    }
}
